src/main/java/picard/illumina/SampleInputParameters.java

```diff
@@ -0,0 +1,119 @@
+package picard.illumina;
+
+import htsjdk.samtools.util.StringUtil;
+import htsjdk.samtools.util.Tuple;
+import picard.illumina.parser.ReadDescriptor;
+import picard.illumina.parser.ReadStructure;
+import picard.illumina.parser.ReadType;
+import picard.util.IlluminaUtil;
+import picard.util.TabbedTextFileWithHeaderParser;
+
+import java.io.File;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * This class encloses sample input parameters used for any ExtractBarcodesProgram (ExtractIlluminaBarcodes,
+ * IlluminaBasecallsToFastq, IlluminaBasecallsToSam). This is used to parse tab delimited sample information including
+ * barcodes, library names and output prefixes.
+ */
+public class SampleInputParameters {
+    /**
+     * Column header for the first barcode sequence (preferred).
+     */
+
+    public static final Set<String> BARCODE_PREFIXES = new HashSet<>(
+            Arrays.asList(SampleInputParameters.BARCODE_SEQUENCE_COLUMN, SampleInputParameters.BARCODE_COLUMN)
+    );
+
+    public static final String BARCODE_COLUMN = "barcode";
+    public static final String BARCODE_SEQUENCE_COLUMN = "barcode_sequence";
+
+    /**
+     * Column header for the barcode name.
+     */
+    public static final String BARCODE_NAME_COLUMN = "barcode_name";
+    /**
+     * Column header for the library name.
+     */
+    public static final String LIBRARY_NAME_COLUMN = "library_name";
+
+
+    /**
+     * Parses any one of the following types of files:
+     *
+     * ExtractIlluminaBarcodes      BARCODE_FILE
+     * IlluminaBasecallsToFastq     MULTIPLEX_PARAMS
+     * IlluminaBasecallsToSam       LIBRARY_PARAMS
+     *
+     * This will validate to file format as well as populate a Map of barcodes to metrics.
+     *
+     * @param inputFile         The input file that is being parsed
+     * @param readStructure     The read structure for the reads of the run
+     */
+    protected static Tuple<Map<String, BarcodeMetric>, List<String>> parseInputFile(final File inputFile,
+                                          final ReadStructure readStructure) {
+        List<String> messages = new ArrayList<>();
+        Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
+        try (final TabbedTextFileWithHeaderParser barcodesParser = new TabbedTextFileWithHeaderParser(inputFile)) {
+            List<String> validBarcodeColumns = barcodesParser.columnLabels().stream().filter(name -> {
+                boolean isValidPrefix = false;
+                for (String columnPrefix : BARCODE_PREFIXES) {
+                    isValidPrefix |= name.toUpperCase().startsWith(columnPrefix.toUpperCase());
+                }
+                return isValidPrefix;
+            }).collect(Collectors.toList());
```

Comment:
I think I would a) extract some of this logic into a function called `getNameOfFirstBarcodeColumn()` that looks something like:

```java
final String getNameOfFirstBarcodeColumn(final Set<String> cols) {
   for (final String prefix : BARCODE_PREFIXES) {
        if (cols.contains(prefix)) return prefix;
        if (cols.contains(prefix + "_1") return prefix + "_1;
    }

    throw new IllegalStateException("Couldn't find column containing first barcode sequence. Expected ...");
}
```

then b) in the loop below use the returned value if `barcodeNum = 0` else use the prefix plus the `_barcodeNum`

Reply:
Yeah its a bit more annoying than that since a single barcode is just `BARCODE` and multiple would be `BARCODE_1` `BARCODE_2` so depending on how many valid columns you have you'd either be appending the numbered suffix or not. 

I'll try to rework this though.